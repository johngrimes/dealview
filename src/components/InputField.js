import React from 'react'

class InputField extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      value: props.value,
      touched: false,
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
  }

  handleChange(event) {
    const target = event.target
    this.setState(
      () => ({ value: target.value, touched: true }),
      () => {
        if (this.props.onChange) this.props.onChange(target.value)
      }
    )
  }

  handleFocus(event) {
    const target = event.target
    if (this.props.onFocus) {
      this.props.onFocus(target.name)
    }
  }

  setFocus() {
    if (this.props.focus === this.props.name) {
      this.ref.focus()
    }
  }

  componentWillReceiveProps(props) {
    this.setState(() => ({ value: props.value }))
  }

  componentDidMount() {
    this.setFocus()
  }

  componentDidUpdate() {
    this.setFocus()
  }

  render() {
    const { name, type, label, placeholder, min, errors, forceErrorDisplay } = this.props
    const { touched } = this.state
    const value = this.state.value ? this.state.value : ''
    const labelTag = label
      ? <label htmlFor={name}>
          {label}
        </label>
      : null
    const errorTags = []
    if (errors) {
      errors.forEach((msg, i) =>
        errorTags.push(
          <div key={i} className='error'>
            {msg}
          </div>
        )
      )
    }
    const className = touched && errorTags.length > 0 ? 'with-errors' : ''
    const props = {
      name,
      className,
      type,
      placeholder,
      min,
      value,
      onChange: this.handleChange,
      onFocus: this.handleFocus,
      ref: input => (this.ref = input),
    }

    return (
      <div className='control-group'>
        {labelTag}
        <input {...props} />
        {(forceErrorDisplay || touched) &&
          errorTags.length > 0 &&
          <div className='errors'>
            {errorTags}
          </div>}
      </div>
    )
  }
}

export default InputField
