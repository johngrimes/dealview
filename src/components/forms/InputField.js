// @flow

import React from 'react'

import type { FieldErrors } from '../../utils/FormValidation.js'

type Props = {
  name: string,
  type: string,
  value: string,
  label?: string,
  placeholder?: string,
  errors?: FieldErrors,
  forceErrorDisplay?: boolean,
  focus?: string,
  onChange?: (value: string) => void,
  onFocus?: (fieldName: string) => void
}

type State = {
  value: string,
  touched: boolean
}

class InputField extends React.Component {
  props: Props
  state: State
  ref: HTMLInputElement
  handleChange: (event: Event) => true
  handleFocus: (event: Event) => true

  constructor(props: Props) {
    super(props)
    this.state = {
      value: props.value,
      touched: false
    }

    this.handleChange = (event) => {
      const target = event.target
      if (target instanceof HTMLInputElement) {
        this.setState(
          () => ({ value: target.value, touched: true }),
          () => { if (this.props.onChange) this.props.onChange(target.value) }
        )
      }
      return true
    }

    this.handleFocus = (event) => {
      const target = event.target
      if (this.props.onFocus && target instanceof HTMLInputElement) {
        this.props.onFocus(target.name)
      }
      return true
    }
  }

  componentWillReceiveProps(props: Props) {
    this.setState(() => ({ value: props.value }))
  }

  setFocus() {
    if (this.props.focus === this.props.name) {
      this.ref.focus()
    }
  }
  componentDidMount() { this.setFocus() }
  componentDidUpdate() { this.setFocus() }

  render() {
    const labelTag = this.props.label ? <label htmlFor={this.props.name}>{this.props.label}</label> : null

    const errorTags = []
    if (this.props.errors) {
      this.props.errors.forEach((msg, i) =>
        errorTags.push(<div key={i} className='error'>{msg}</div>)
      )
    }

    const inputClass = this.state.touched && errorTags.length > 0
      ? 'with-errors'
      : ''

    return (
      <div className='control-group'>
        {labelTag}
        <input name={this.props.name} className={inputClass} type={this.props.type} placeholder={this.props.placeholder} value={this.state.value} onChange={this.handleChange} onFocus={this.handleFocus} ref={input => this.ref = input} />
        {(this.props.forceErrorDisplay || this.state.touched) && errorTags.length > 0 && <div className='errors'>{errorTags}</div>}
      </div>
    )
  }
}

export default InputField
