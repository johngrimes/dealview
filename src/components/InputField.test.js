import React from 'react'
import { shallow, mount } from 'enzyme'

import InputField from './InputField.js'

describe('InputField', () => {
  it('should render', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      value: 'something',
      label: 'Some Input',
      placeholder: 'Type here...',
    }
    const wrapper = shallow(<InputField {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should be tolerant of an undefined value prop', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      label: 'Some Input',
      placeholder: 'Type here...',
    }
    const wrapper = shallow(<InputField {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should display errors if touched', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      errors: [ 'does not look right', 'smells funny' ],
    }
    const wrapper = shallow(<InputField {...props} />)
    wrapper.setState(
      {
        value: 'something',
        touched: true,
      },
      () => {
        expect(wrapper.find('.error')).toHaveLength(2)
        expect(wrapper).toMatchSnapshot()
      }
    )
  })

  it('should not display errors if not touched', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      errors: [ 'does not look right', 'smells funny' ],
    }
    const wrapper = shallow(<InputField {...props} />)
    expect(wrapper.find('.error')).toHaveLength(0)
  })

  it('should display errors if not touched but forceErrorDisplay is true', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      errors: [ 'does not look right', 'smells funny' ],
      forceErrorDisplay: true,
    }
    const wrapper = shallow(<InputField {...props} />)
    expect(wrapper.find('.error')).toHaveLength(2)
  })

  it('should call onChange when the input changes', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      onChange: jest.fn(),
    }
    const wrapper = shallow(<InputField {...props} />)
    wrapper.find('input').prop('onChange')({ target: { value: 'something' } })
    expect(props.onChange).toHaveBeenCalledWith('something', undefined)
  })

  it('should call onFocus when the input is focused', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      onFocus: jest.fn(),
    }
    const wrapper = shallow(<InputField {...props} />)
    wrapper.find('input').prop('onFocus')({ target: { name: 'someInput' } })
    expect(props.onFocus).toHaveBeenCalledWith('someInput')
  })

  it('should focus the input when the focus prop is passed', () => {
    const props = {
      name: 'someInput',
      type: 'text',
    }
    const wrapper = mount(<InputField {...props} />)
    const spy = jest.spyOn(wrapper.find('input').getNode(), 'focus')
    wrapper.setProps({ focus: 'someInput' })
    expect(spy).toHaveBeenCalled()
  })
})
