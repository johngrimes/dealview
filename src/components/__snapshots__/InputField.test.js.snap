// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField should be tolerant of an undefined value prop 1`] = `
<div
  className="control-group"
>
  <label
    htmlFor="someInput"
  >
    Some Input
  </label>
  <input
    name="someInput"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Type here..."
    type="text"
    value=""
  />
</div>
`;

exports[`InputField should display errors if touched 1`] = `
<div
  className="control-group"
>
  <input
    name="someInput"
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
</div>
`;

exports[`InputField should render 1`] = `
<div
  className="control-group"
>
  <label
    htmlFor="someInput"
  >
    Some Input
  </label>
  <input
    name="someInput"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Type here..."
    type="text"
    value="something"
  />
</div>
`;
